pkg/virt-controller/liveupdate/handlers.go

```diff
@@ -0,0 +1,151 @@
+	t := reflect.TypeOf(obj)
+	path = strings.Trim(path, "/")
+
+	for _, field := range strings.Split(path, "/") {
+		elem, found := t.FieldByName(field)
+		if !found {
+			return false
+		}
+
+		t = elem.Type
+
+		if t.Kind() == reflect.Pointer {
+			t = t.Elem()
+		}
+	}
+
+	return true
+}
+
+func getField(vm *v1.VirtualMachine, path string) (reflect.Value, error) {
+	path = strings.Trim(path, "/")
+	v := reflect.ValueOf(vm).Elem()
+
+	for _, field := range strings.Split(path, "/") {
+		v = v.FieldByName(field)
+
+		if v.Kind() == reflect.Pointer {
+			v = v.Elem()
+		}
+
+		if !v.IsValid() {
+			return reflect.Value{}, fmt.Errorf("field '%s' not found", path)
+		}
+	}
+	return v, nil
+}
+
+func copyFields(dstVM, srcVM *v1.VirtualMachine, pathList []string) error {
+	for _, path := range pathList {
+		value, err := getField(srcVM, path)
+		if err != nil {
+			return err
+		}
+		dst, err := getField(dstVM, path)
+		if err != nil {
+			return err
+		}
+
+		if !dst.CanSet() {
+			return fmt.Errorf("field '%s' of destination object cannot be set", path)
+		}
+		dst.Set(value)
+	}
+	return nil
+}
+
+type LiveUpdateHandler interface {
+	// GetManagedFields() should return an array containing the paths
+	// that must be filtered out when comparing a new VM spec
+	// with its last seen version.
+	//
+	// Every string in an array must be a forward-slash separated
+	// path of the field that has to be filtered out.
+	// Example: to filter out .spec.template.spec.domain.memory.guest you should supply
+	// '/Spec/Template/Spec/Domain/Memory/Guest'.
+	// Every element of the path is a Go struct field name.
+	GetManagedFields() []string
+	// HandleLiveUpdate() should implement the live-update operation
+	// the function receives the current VM and VMI objects as parameters
+	HandleLiveUpdate(vm *v1.VirtualMachine, vmi *v1.VirtualMachineInstance) error
+}
+
+type LiveUpdater struct {
+	handlers []LiveUpdateHandler
+}
+
+func (up *LiveUpdater) RegisterHandlers(handlers ...LiveUpdateHandler) error {
+	for _, handler := range handlers {
+		for _, field := range handler.GetManagedFields() {
+			if !fieldExists(v1.VirtualMachine{}, field) {
+				return fmt.Errorf("cannot register live-update handler, field '%s' does not exist", field)
+			}
+		}
+
+		up.handlers = append(up.handlers, handler)
+	}
+	return nil
+}
+
+func (up *LiveUpdater) FilterUpdatableFields(vm, oldVM *v1.VirtualMachine) error {
+	for _, handler := range up.handlers {
+		if err := copyFields(vm, oldVM, handler.GetManagedFields()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (up *LiveUpdater) HandleLiveUpdates(vm, oldVM *v1.VirtualMachine, vmi *v1.VirtualMachineInstance) error {
+	for _, handler := range up.handlers {
+		for _, field := range handler.GetManagedFields() {
+			oldValue, err := getField(oldVM, field)
+			if err != nil {
+				return err
+			}
+
+			newValue, err := getField(vm, field)
+			if err != nil {
+				return err
+			}
```

Comment:
Actually, the same comment https://github.com/kubevirt/kubevirt/pull/11748/files#r1572337056 is true here. For the volumes we might want to filter certain volumes types. An the getField for new and old values isn't the same.

Reply:
We would need to cover cases like the one described in https://github.com/kubevirt/kubevirt/pull/11700#discussion_r1572173667 and this requires the old and new version of the VM.

Reply:
A version without reflection that allows the type implementing the interface to do this was already in the works.